File: onyx-database/src/main/java/com/onyx/entity/AbstractSystemEntity.java
package com.onyx.entity;

import com.onyx.persistence.ManagedEntity;
import com.onyx.persistence.annotations.Attribute;
import com.onyx.persistence.annotations.PreInsert;
import com.onyx.persistence.annotations.PreUpdate;

import java.util.Date;

/**
 * Created by timothy.osborn on 3/15/15.
 *
 * Contains date modified information
 */
class AbstractSystemEntity extends ManagedEntity
{

    @Attribute
    private Date dateUpdated;

    @Attribute
    private Date dateCreated;

    @SuppressWarnings("unused")
    public Date getDateUpdated()
    {
        return dateUpdated;
    }

    @SuppressWarnings("unused")
    public void setDateUpdated(Date dateUpdated)
    {
        this.dateUpdated = dateUpdated;
    }

    @SuppressWarnings("unused")
    public Date getDateCreated()
    {
        return dateCreated;
    }

    @SuppressWarnings("unused")
    public void setDateCreated(Date dateCreated)
    {
        this.dateCreated = dateCreated;
    }

    @PreInsert
    @SuppressWarnings("unused")
    protected void onPrePersist()
    {
        dateCreated = new Date();
    }

    @PreUpdate
    @SuppressWarnings("unused")
    protected void onPreUpdate()
    {
        dateUpdated = new Date();
    }
}

File: onyx-database/src/main/kotlin/com/onyx/entity/AbstractSystemEntity.kt
package com.onyx.entity

import com.onyx.persistence.ManagedEntity
import com.onyx.persistence.annotations.Attribute
import com.onyx.persistence.annotations.PreInsert
import com.onyx.persistence.annotations.PreUpdate
import java.util.*

/**
 * Created by timothy.osborn on 3/15/15.
 *
 * Contains date modified information
 */
@Suppress("MemberVisibilityCanPrivate")
abstract class AbstractSystemEntity : ManagedEntity() {

    @Attribute
    var dateUpdated: Date? = null

    @Attribute
    var dateCreated: Date? = null

    @PreInsert
    @Suppress("UNUSED")
    private fun onPrePersist() {
        dateCreated = Date()
    }

    @PreUpdate
    @Suppress("UNUSED")
    private fun onPreUpdate() {
        dateUpdated = Date()
    }
}
